### DISCUSSION TITLE
SUMMA parallelization

### DISCUSSION THREAD

**Discussion by AshlinAnn** [CH-Earth/summa#493] [Q&A] [answered]
Hi, I am trying to set up SUMMA on Cray XC40. I am using a gfortran compiler. How do I enable parallelization for SUMMA? What changes should I bring in Makefile?

Thanks.

**Comment by wknoben** [comment, marked as answer]
Hi Ashlin, 
Currently the most resource-efficient way to parallelize SUMMA runs is to use the built-in `-g` option. This option is used as follows:
```
startGRU = 1  # index of the GRU in the input .nc files; note that this refers to the index in the file (1..n) and NOT to the gruId variable
numGRU = 15   # number of GRUs to run, starting with startGRU
summa.exe -g startGRU numGRU -m filemanager.txt # runs SUMMA for 15 GRUs, starting with the first GRU in file
```
This will result in a split-domain output file, containing your requested outputs only for these 15 GRUs. The GRU indices for which the file has simulations are automatically included in the file's name.

You can access a brief description of what this does through `summa.exe --help`

**Comment by NicWayand** [comment]
Are there some more best practices for running SUMMA at scale (millions of points)?

**Comment by andywood** [comment]
Hi Nic, I typically use split domain runs with empirical load balancing of
GRU ranges (ie with -g) across cores.  It's clunky but works fine.  We do a
preamble run to determine the rough GRU ranges, then fix them in a joblist
which can be run over many nodes.  The new 'agents' stuff being built now
may be a bit more streamlined and/or effective, but may also come with
extra dependencies and overhead.  It's probably worth benchmarking the
different approaches for performance.  I'd guess that w/o a formal MPI
approach and more code optimization (ie parallel I/O & data objects) that
we're reasonably close to what is achievable, efficiency-wise, versus a
rudimentary domain-split.

**Reply by NicWayand** [reply]
I see there is a compile option with OpenMI, is that option working?

**Comment by NicWayand** [comment]
Thank you Andy. I will try the split domain approach for now.

**Comment by andywood** [comment]
apparently openMP works but doesn't scale all that well.  I've never tried
it versus the other more manual parallelization approaches, since I need to
do those anyway with multi-node jobs
